Documentation/guides/internals/debug-jni-objrefs.md

```diff
@@ -0,0 +1,400 @@
+```
+
+The `PeerInfo.State` value of `Collected` means that the instance has been collected.
+
+### Dalvik Abort Crash Logs
+
+Consider the [crash via Dalvik Abort](#crash-dalvik-abort):
+
+```
+java_vm_ext.cc:579] JNI DETECTED ERROR IN APPLICATION: JNI ERROR (app bug): jobject is an invalid local reference: 0x75 (deleted reference at index 7 in a table of size 7)
+```
+
+Of note is that the abort message mentions *local*, in "an invalid local reference".  This is a sign that we need to collect the JNI Local Reference logs:
+
+```sh
+adb shell setprop debug.mono.log lref,gref
+
+# run the app…
+
+adb shell run-as @PACKAGE-NAME@ cat files/.__override__/grefs.txt > grefs.txt
+adb shell run-as @PACKAGE-NAME@ cat files/.__override__/lrefs.txt > lrefs.txt
+```
+
+After collecting the JNI Local Reference logs, we search backward, from the end, for the values
+`0x75`.  (If using `less`, search backwards `?` for the value `0x75`.)
+This lands us on:
+
+```
+-l- lrefc 0 handle 0x75/L from thread '(null)'(1)
+   at Android.Runtime.AndroidObjectReferenceManager.DeleteLocalReference(JniObjectReference& value, Int32& localReferenceCount) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/AndroidRuntime.cs:line 140
+   at Java.Interop.JniRuntime.JniObjectReferenceManager.DeleteLocalReference(JniEnvironmentInfo environment, JniObjectReference& reference) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniRuntime.JniObjectReferenceManager.cs:line 70
+   at Java.Interop.JniObjectReference.Dispose(JniObjectReference& reference) in /Users/runner/work/1/s/xamarin-android/external/Java.Interop/src/Java.Interop/Java.Interop/JniObjectReference.cs:line 195
+   at Scratch.GrefCrashNet6.MainActivity.UseInvalidJniHandle() in …/Scratch.GrefCrashNet6/Scratch.GrefCrashNet6/MainActivity.cs:line 24
+   at Scratch.GrefCrashNet6.MainActivity.OnCreate(Bundle savedInstanceState) in …/Scratch.GrefCrashNet6/Scratch.GrefCrashNet6/MainActivity.cs:line 15
+   at Android.App.Activity.n_OnCreate_Landroid_os_Bundle_(IntPtr jnienv, IntPtr native__this, IntPtr native_savedInstanceState) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net6.0/android-31/mcw/Android.App.Activity.cs:line 2781
+   at Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V(_JniMarshal_PPL_V callback, IntPtr jnienv, IntPtr klazz, IntPtr p0) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:line 121
+```
+
+This tells us that the JNI Local Reference was in fact deleted, from `UseInvalidJniHandle()`,
+in `MainActivity.cs:15`.  From there we can investigate our code and fix it so that an
+invalid JNI handle is no longer used.
+
+
+<a name="logcat-parse"></a>
+
+## `logcat-parse`
+
+`logcat-parse` combines an API around the JNI Global Reference logs with a C# REPL.
+It is included with Classic Xamarin.Android installs.
```

Comment:
Do we need to file an issue to distribute this as a .NET global tool? On NuGet.org? Then you could do `dotnet tool install --global logcat-parse`.

Reply:
Yes, I should do that.

Reply:
Might also take the opportunity to rename `logcat-parse`, since I don't think it's parsing `logcat`?  😁

Reply:
It *can* parse `adb logcat` output.

I'm open to alternative names.  `objref-parse`?